code/datums/components/nanites.dm

```diff
@@ -0,0 +1,163 @@
+/datum/component/nanites
+	dupe_mode = COMPONENT_DUPE_UNIQUE_PASSARGS
+
+	var/mob/living/host_mob
+	var/nanite_volume = 100		//amount of nanites in the system, used as fuel for nanite programs
+	var/max_nanites = 500		//maximum amount of nanites in the system
+	var/regen_rate = 0.5		//nanites generated per second
+	var/safety_threshold = 50	//how low nanites will get before they stop processing/triggering
+	var/cloud_id = 0 			//0 if not connected to the cloud, 1-100 to set a determined cloud backup to draw from
+	var/next_sync = 0
+	var/list/datum/nanite_program/programs = list()
+	var/max_programs = 15
+
+	var/stealth = FALSE //if TRUE, does not appear on HUDs and health scans, and does not display the program list on nanite scans
+
+/datum/component/nanites/Initialize(amount)
+	nanite_volume = amount
+
+	//Nanites without hosts are non-interactive through normal means
+	if(isliving(parent))
+		host_mob = parent
+
+		if(!(MOB_ORGANIC in host_mob.mob_biotypes) && !(MOB_UNDEAD in host_mob.mob_biotypes)) //Shouldn't happen, but this avoids HUD runtimes in case a silicon gets them somehow.
+			qdel(src)
+
+		host_mob.hud_set_nanite_indicator()
+		START_PROCESSING(SSprocessing, src)
+		RegisterSignal(host_mob, COMSIG_ATOM_EMP_ACT, .proc/on_emp)
+		RegisterSignal(host_mob, COMSIG_MOB_DEATH, .proc/on_death)
+		RegisterSignal(host_mob, COMSIG_LIVING_ELECTROCUTE_ACT, .proc/on_shock)
+		RegisterSignal(host_mob, COMSIG_LIVING_MINOR_SHOCK, .proc/on_minor_shock)
+		RegisterSignal(host_mob, COMSIG_MOVABLE_HEAR, .proc/on_hear)
+		RegisterSignal(host_mob, COMSIG_NANITE_SIGNAL, .proc/receive_signal)
+		if(cloud_id)
+			cloud_sync()
+
+/datum/component/nanites/Destroy()
+	STOP_PROCESSING(SSprocessing, src)
+	for(var/X in programs)
+		qdel(X)
+	if(host_mob)
+		host_mob.hud_set_nanite_indicator()
+	return ..()
+
+/datum/component/nanites/InheritComponent(datum/component/new_nanites, i_am_original, list/arguments)
+	adjust_nanites(arguments[1]) //just add to the nanite volume
+
+/datum/component/nanites/process()
+	if(nanite_volume <= 0) //oops we ran out
+		qdel(src)
+	if(host_mob)
+		if(!(MOB_ORGANIC in host_mob.mob_biotypes) && !(MOB_UNDEAD in host_mob.mob_biotypes))
```

Comment:
You need to detect when the change as it happens, not in process.

Reply:
Time for a species change signal! That one will see some use, i'd bet